### PR TITLE
GT_STORE_BLK - do not call memset for blocks containg gc pointers on heap

### PR DESCRIPTION
**Pull request by EgorBo**
This PR applies fix suggested by @jkotas to always inline "memset" for structs with GC references since `JIT_Memset` doesn't provide us with needed atomicity guarantees. An alternative solution is to introduce a new memset helper (it requires R2R format update).
1) Current PR still allows Memset when target is a local (= not visible) - for that we're also allowed to use SIMD (current Main behaviour) - is it correct?
2) It seems that we should allow SIMD as well if we manually align it to 16 bytes (and put it in nongc zone) to optimize the long sequence of 8-byte moves for large structs

```cs
void Test(ref LargeStructWithGC s)
{
    s = default;
}

struct LargeStructWithGC // 184 bytes
{
    long b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, r, s, t, u, v, w, z;
    string a;
}
```
Was:
```asm
; Method Program:Test(byref):this (FullOpts)
       sub      rsp, 40
       mov      rcx, rdx
       xor      edx, edx
       mov      r8d, 184
       call     CORINFO_HELP_MEMSET
       nop      
       add      rsp, 40
       ret      
; Total bytes of code: 26
```
Now:
```asm
; Method StructWithGC_Zeroing:Test(byref):this (FullOpts)
       xor      eax, eax
       mov      qword ptr [rdx], rax
       mov      ecx, 176 ;;; 184 - 8
G_M2317_IG03:
       mov      qword ptr [rdx+rcx], rax
       sub      rcx, 8
       jne      SHORT G_M2317_IG03
       ret      
; Total bytes of code: 21
```

### CODE REVIEW

**Comment by jkotas** [comment]
> Current PR still allows Memset when target is a local (= not visible) - for that we're also allowed to use SIMD (current Main behaviour) - is it correct?

Yes, that sounds right.

**Comment by jakobbotsch** [comment]
I think some fallback for large structs will be needed, either as an inline loop or helper callback, otherwise we can hit encoder limitations. One of the first lessons I learned working on the JIT in #61521.

**Comment by EgorBo** [comment]
> I think some fallback for large structs will be needed, either as an inline loop or helper callback, otherwise we can hit encoder limitations. One of the first lessons I learned working on the JIT in #61521.

Yeah I already hit it locally (added more tests)

**Comment by EgorBo** [comment]
For reference, a corresponding fix to introduce a new helper (with r2r format bump): https://github.com/EgorBo/runtime-1/commit/262c43c4a22ead56d23ca44ace37c961ca613145

**Comment by jkotas** [comment]
It is why we have `ZeroMemoryInGCHeap`

**Comment by EgorBo** [comment]
PTAL @dotnet/jit-contrib @jkotas 

I implemented the loop (for cases when size is too big to be unrolled). [Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=487372&view=ms.vss-build-web.run-extensions-tab) are not too big and don't touch sensitive stuff it seems (except, maybe ValueTasks with large state). The loop can be additionally optimized further if needed.

An alternative solution is to introduce a new helper that will `ZeroMemoryInGCHeap` (something like https://github.com/EgorBo/runtime-1/commit/262c43c4a22ead56d23ca44ace37c961ca613145 + needs a few more changes for X86_32) - but that requires R2R format bump that might not be desired for backports.

File: src/coreclr/jit/codegenxarch.cpp
Comment (BruceForstall) [inline src/coreclr/jit/codegenxarch.cpp:3353]: Should this use `GetSingleTempReg` instead?

File: src/coreclr/jit/codegenxarch.cpp
Comment (BruceForstall) [inline src/coreclr/jit/codegenxarch.cpp:3344]: Couldn't you get rid of the initial `mov qword ptr [dstReg], zeroReg` by using `jns` instead of `jne`?

e.g., https://godbolt.org/z/Y7bEGzrfx

File: src/coreclr/jit/codegenarmarch.cpp
Comment (BruceForstall) [inline src/coreclr/jit/codegenarmarch.cpp:3300]: Not true: `sub` has a `subs` variant, using `INS_FLAGS_SET`, e.g.,
```
GetEmitter()->emitIns_R_I(INS_sub, EA_PTRSIZE, offsetReg, TARGET_POINTER_SIZE, INS_FLAGS_SET);
```

**Review by BruceForstall — CHANGES_REQUESTED** [review]
You should add a section to https://github.com/dotnet/runtime/blob/main/docs/design/coreclr/botr/clr-abi.md detailing the GC requirements for memset/memcpy atomicity.

File: src/coreclr/jit/codegenxarch.cpp
Comment (EgorBo) [reply src/coreclr/jit/codegenxarch.cpp:3344]: I think it serves as a nullcheck here, if we don't have it then for some large size we'll start our "reversed" zeroing from `null + largeConstant` and may not be able to handle that as NRE? (will be AVE instead)

**Comment by EgorBo** [comment]
Another problem we discovered (with @MichalPetryka) is that RISC-V (and LA64?) don't have addressing modes for indices (base + index), only for constant imms. It means we need to prepare the address in a separate temp reg so I should add it to gcinfo or mark loop as nogc I assume.

**Comment by am11** [comment]
> RISC-V (and LA64?) don't have addressing modes for indices (base + index)

Yup, `emitInsLoadStoreOp` in `emit{loong,riscv}64.cpp` also handle them as complex addressing modes.

File: src/coreclr/jit/codegenxarch.cpp
Comment (BruceForstall) [reply src/coreclr/jit/codegenxarch.cpp:3344]: If it is a required null check, make sure there is a comment here that makes that clear. (And if there are any conditions under which the null check can be omitted?)

Comment (EgorBo) [reply src/coreclr/jit/codegenxarch.cpp:3344]: Added a comment, we probably can detect cases where we don't need the nullcheck, but it also does the useful work - it clears the first pointer (the loop doesn't clear it again due to `JUMP_NE`) so it's probably not worth the effort to try to skip it since diffs aren't too big and, presumably, we only emit loops for cases when we don't unroll - large sizes, so most of the performance is spent inside the loop.

File: src/coreclr/jit/codegenxarch.cpp
Comment (EgorBo) [reply src/coreclr/jit/codegenxarch.cpp]: Changed

File: src/coreclr/jit/codegenarmarch.cpp
Comment (EgorBo) [reply src/coreclr/jit/codegenarmarch.cpp]: Applied

File: src/coreclr/jit/codegenarmarch.cpp
Comment (BruceForstall) [inline src/coreclr/jit/codegenarmarch.cpp:3265]: Are you going to address this "TODO" before merging? It seems like we want to fully implement this fix in one change if we intend to back-port it.

File: src/coreclr/jit/lowerloongarch64.cpp
Comment (BruceForstall) [inline src/coreclr/jit/lowerloongarch64.cpp:332]: It looks like the codegenloongarch64.cpp code is unfinished (with "TODO"). Should this change be commented out (left as it was before), and open an issue to fix the codegen and enable it?

File: src/coreclr/jit/lowerriscv64.cpp
Comment (BruceForstall) [inline src/coreclr/jit/lowerriscv64.cpp:280]: It looks like the codegenriscv64.cpp code is unfinished (with "TODO"). Should this change be commented out (left as it was before), and open an issue to fix the codegen and enable it?

File: src/coreclr/jit/codegenarmarch.cpp
Comment (EgorBo) [reply src/coreclr/jit/codegenarmarch.cpp:3265]: Yes, just wanted to get a review on the rest so then I can involve the riscv/la folks to help me/review these parts

Comment (BruceForstall) [reply src/coreclr/jit/codegenarmarch.cpp:3265]: The non-riscv/la64 looks good to me.

If you want to wait for riscv/la64 help, that's ok. But as I indicated in my other comments, I think it would be fine to just not enable the new unroll mode for those architectures and let those architecture owners do it.

**Comment by EgorBo** [comment]
@BruceForstall a question regarding the

It seems that I mess with the GC info when I create a BB for the loop so the byref (address of the struct) is no longer reported?
I added `GetEmitter()->emitDisableGC();` and `GetEmitter()->emitEnableGC();` for it.

```cs
static void Test(ref LargeStructWithGC s)
{
    s = default;
}
```
Codegen diff (with `DOTNET_JitDisasmWithGC=1`): https://www.diffchecker.com/1sTjo60b/

is this correct? cc @dotnet/jit-contrib
or should I some how fix the gcinfo instead?

**Comment by BruceForstall** [comment]
I don't see why you need to disable GC for this loop.

I do see this at the top of the loop:
```
; byrRegs -[rcx]
```

Why is this getting killed? It looks like the call to `genDefineTempLabel` must be getting called with an empty byref set. why is this? Has the liveness update for GT_STORE_BLK already killed rcx? Do you need LSRA to extend the lifetime or something similar?

**Comment by SingleAccretion** [comment]
Liveness of the destination pointer probably needs to be manually extended with `gcInfo.gcMarkRegPtrVal`, like here:

https://github.com/dotnet/runtime/blob/1b3208d859e20fb37918a11c7e9d2554326f5b67/src/coreclr/jit/codegenxarch.cpp#L5259-L5265

File: src/coreclr/jit/codegenriscv64.cpp
Comment (MichalPetryka) [inline src/coreclr/jit/codegenriscv64.cpp:6112]: `tempReg` also needs to be marked with gcMarkRegPtrVal.

File: src/coreclr/jit/lowerxarch.cpp
Comment (BruceForstall) [inline src/coreclr/jit/lowerxarch.cpp:405]: The "too big to unroll" case is expected to be rare, right?

Should `getUnrollThreshold` be artificially reduced under stress to make the loop case happen more frequently?

And/or should we go down the BlkOpKindLoop even for non-GC-pointer-on-heap zeroing under some stress percentage?

Basically, do we get enough test coverage as-is, or should we artificially adjust test coverage sometimes under stress?

**Review by gbalykov — APPROVED** [review]
RISC-V changes look good to me. Thanks! 

> From my understanding, RISC-V doesn't have stores with "base + index" addressing mode, am I right?

Yes, you're right. As far as I know the only extension that proposed to include register-indexed addressing mode was recent `Znew`, nothing ratified yet, and probably won't be in near future too.

> it's probably possible to rewrite it to use just a single temp reg but I am not sure how perf sensitive this code is

I think we can keep it as a todo for now.

File: src/coreclr/jit/codegenloongarch64.cpp
Comment (MichalPetryka) [inline src/coreclr/jit/codegenloongarch64.cpp:6438]:
```suggestion
        // *(dstReg + offset) = 0
        GetEmitter()->emitIns_R_R_R(INS_stx_d, EA_PTRSIZE, zeroReg, dstReg, offsetReg);
```

Comment (EgorBo) [reply src/coreclr/jit/codegenloongarch64.cpp]: Thanks! Applied locally since it needed more changes

File: src/coreclr/jit/codegenloongarch64.cpp
Comment (clamp03) [reply src/coreclr/jit/codegenloongarch64.cpp:6438]: @EgorBo In RISCV64, `REG_R0` is always const zero too. :)

**Review by clamp03 — APPROVED** [review]
LGTM Thank you @EgorBo 
cc @tomeksowi

**Review by tomeksowi — APPROVED** [review]
RISC-V part LGTM, thx @EgorBo

File: src/coreclr/jit/codegenloongarch64.cpp
Comment (EgorBo) [reply src/coreclr/jit/codegenloongarch64.cpp]: Thanks! Added `REG_R0` to both archs.

> Maybe we can optimize these by avoiding the computing the TempLabel offset.

@shushanhf could you please clarify what you mean?

**Comment by EgorBo** [comment]
@BruceForstall can you please review & approve if it's fine?

File: src/coreclr/jit/codegenarmarch.cpp
Comment (BruceForstall) [inline src/coreclr/jit/codegenarmarch.cpp:3268]: Can you remove this TODO now?

File: src/coreclr/jit/codegenarmarch.cpp
Comment (BruceForstall) [inline src/coreclr/jit/codegenarmarch.cpp:3276]: This is confusing because you would never "mov zeroReg, wzr".

**Review by BruceForstall — COMMENTED** [review]
LGTM. Minor notes on comments.

File: src/coreclr/jit/codegenriscv64.cpp
Comment (t-mustafin) [inline src/coreclr/jit/codegenriscv64.cpp:6128]: Think one instruction could be optimized out of this loop:
```c++
        const regNumber tempReg   = initBlkNode->ExtractTempReg();
        instGen_Set_Reg_To_Imm(EA_PTRSIZE, tempReg, size - TARGET_POINTER_SIZE);

        // tempReg = dstReg + tempReg (a new interior pointer, but in a nongc region)
        GetEmitter()->emitIns_R_R_R(INS_add, EA_PTRSIZE, tempReg, dstReg, tempReg);

        BasicBlock* loop = genCreateTempLabel();
        genDefineTempLabel(loop);
        GetEmitter()->emitDisableGC(); // TODO: add gcinfo to tempReg and remove nogc

        // *tempReg = 0
        GetEmitter()->emitIns_R_R_I(INS_sd, EA_PTRSIZE, REG_R0, tempReg, 0);
        // tempReg = tempReg - 8
        GetEmitter()->emitIns_R_R_I(INS_addi, EA_PTRSIZE, tempReg, tempReg, -8);
        // if (tempReg != dstReg) goto loop;
        GetEmitter()->emitIns_J(INS_bne, loop, (int)tempReg | ((int)dstReg << 5));
```

File: src/coreclr/jit/codegenloongarch64.cpp
Comment (shushanhf) [reply src/coreclr/jit/codegenloongarch64.cpp]: The defination of `BasicBlock* loop = genCreateTempLabel();` will need the computation the offset of the branch target which compuated by the emitBinding within the a loop.

But here is the loop target's offset is const, so we can emit the branch directly without computing within the emitBinding.

File: src/coreclr/jit/codegenriscv64.cpp
Comment (am11) [reply src/coreclr/jit/codegenriscv64.cpp:6128]: Related to RISC-V memcpy:
* https://www.phoronix.com/news/LLVM-libc-Faster-memcpy
* https://news.ycombinator.com/item?id=36029262

Maybe we can try to bring it closer to it in the future when vector ext. work would begin. :)

File: src/coreclr/jit/codegenloongarch64.cpp
Comment (shushanhf) [reply src/coreclr/jit/codegenloongarch64.cpp]: @EgorBo 
As you had updated the `zeroReg` as `REG_R0`, we can optimize liking these:
```cpp
        // loop begin:
        // *(dstReg + offsetReg) = 0
        GetEmitter()->emitIns_R_R_R(INS_stx_d, EA_PTRSIZE, zeroReg, dstReg, offsetReg);
        // offsetReg = offsetReg - 8
        GetEmitter()->emitIns_R_R_I(INS_addi_d, EA_PTRSIZE, offsetReg, offsetReg, -8);
        // if (offsetReg != 0) goto loop;
        GetEmitter()->emitIns_R_I(INS_bnez, EA_8BYTE, offsetReg, -2 << 2);
```

Note:
The ins `INS_beq` of the `        GetEmitter()->emitIns_J_cond_la(INS_beq, loop, offsetReg, zeroReg);` is error,
My previous suggestion is `        GetEmitter()->emitIns_R_I(INS_bnez, EA_8BYTE, offsetReg, -2 << 2);`  or
`        GetEmitter()->emitIns_R_R_I(INS_bne, EA_8BYTE, offsetReg, zeroReg, -2 << 2);`

And here emitting a branch instruction directly will avoid the compuatation within `GetEmitter()->emitIns_J_cond_la();`,
this is also applicable to the other Archectures, especially for the RISC CPUs liking LoongArch, ARM, RISC-V.

**Review by BruceForstall — APPROVED** [review]

File: src/coreclr/jit/codegenloongarch64.cpp
Comment (shushanhf) [inline src/coreclr/jit/codegenloongarch64.cpp:6433]:
```suggestion
        // loop begin:
        // *(dstReg + offsetReg) = 0
        GetEmitter()->emitIns_R_R_R(INS_stx_d, EA_PTRSIZE, zeroReg, dstReg, offsetReg);
        // offsetReg = offsetReg - 8
        GetEmitter()->emitIns_R_R_I(INS_addi_d, EA_PTRSIZE, offsetReg, offsetReg, -8);
        // if (offsetReg != 0) goto loop;
        GetEmitter()->emitIns_R_I(INS_bnez, EA_8BYTE, offsetReg, -2 << 2);
```

As the comment https://github.com/dotnet/runtime/pull/95530#discussion_r1421867680

**Review by shushanhf — CHANGES_REQUESTED** [review]

File: src/coreclr/jit/codegenloongarch64.cpp
Comment (EgorBo) [reply src/coreclr/jit/codegenloongarch64.cpp]: Thanks!!

File: src/coreclr/jit/codegenriscv64.cpp
Comment (EgorBo) [reply src/coreclr/jit/codegenriscv64.cpp]: Applied, thanks!

**Comment by EgorBo** [comment]
Since the last approval there only changes were in RISC/LA to address feedback from RISC/LA teams.

**Comment by EgorBo** [comment]
@jkotas does this need to be backported to any release?

Presumably the actual bug was there since .NET Framework?

**Comment by jkotas** [comment]
> Presumably the actual bug was there since .NET Framework?

.NET Framework has a different (slower) implementation of JIT_MemSet that does not appear to have the tearing problem. It does not use `rep stosb`.

> does this need to be backported to any release?

It was found via internal escalation (service crashing 3x per day). We should backport it to .NET 8.  

(It was hit on .NET 6, but the affected customer is planning to move to .NET 8 soon. .NET 8 backport should be good enough.)

**Pull request merged**
